### PR TITLE
Add back: "functionalization: add support for zero_()"

### PR DESCRIPTION
**Pull request by bdhirsh**
Adding back `at::zero()`.

The original plan was that `zero()` would decompose into `zeros_like()`, that way anyone performing tracing would see `zeros_like()` in their graph, and don't have to worry about the new `zero()` op (which is not exposed to python).

I'm pretty sure that we're forced to let `zero()` show up directly in the trace though, because its behavior w.r.t autograd is different compared to `zeros_like()`.

```
# zeros_like() ignores autograd on the input when generating the output
>>> a = torch.ones(2, requires_grad=True)
>>> b = a + a
>>> c = torch.zeros_like(b)
>>> c.requires_grad
False
>>> c.is_leaf
True

# zero_() doesn't ignore autograd (b will still have a grad_fn after being mutated)
# so zero() should do the same.
>>> a = torch.ones(2, requires_grad=True)
>>> b = a + a
>>> b.zero_()
tensor([0., 0.], grad_fn=<ZeroBackward0>)
```

I originally tried to let `zero()` decompose into `zeros_like()` and add a derivative formula for `zeros_like()` to ensure that it shows up as a primitive when tracing. But you can't do that without changing `zeros_like()`'s autograd behavior.

(Specifically, this broken some torchvision code that ran `torch.zeros_like(a, dtype=torch.bool)` under `jit.script`. Before, that code would return a non-autograd tensor. After, it would error since `bool` doesn't work with autograd).

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#76377 Add back: "functionalization: add support for zero_()"**
* #76375 Revert "functionalization: add support for zero_()"

This reverts commit f936c2bc509c2a168c51b50e88edd2fb41adefaf.

### CODE REVIEW

**Comment by ezyang** [comment]
Doesn't this suggest that `zero_` autograd handling is in the wrong here? This is the weird case where you've completely ignored the input tensor contents, which means that technically you can just stop tracking autograd at that point, because nothing you do will propagate. This suggests that `x.zero_()` should flip x from `requires_grad=True` to `requires_grad=False`. @albanD does this seem crazy to you?

I'm going to give you an approve anyway to unblock.

File: tools/autograd/load_derivatives.py [hunk 1]
```diff
@@ -130,8 +130,8 @@ def load_derivatives(
 
 
 @with_native_function
-def cpp_arguments(f: NativeFunction) -> Sequence[Binding]:
-    return CppSignatureGroup.from_native_function(f, method=False).signature.arguments()
+def dispatcher_arguments(f: NativeFunction) -> Sequence[Binding]:
+    return dispatcher.arguments(f.func)
```
Comment (ezyang) [inline tools/autograd/load_derivatives.py:134]: this logic was just wrong before, is that right?

**Review by ezyang — APPROVED** [review]

File: tools/autograd/load_derivatives.py (hunk 1 above)
Comment (bdhirsh) [reply tools/autograd/load_derivatives.py:134]: yeah (it would silently try to spit out `TensorOptions` arguments instead of their exploded version, although that doesn't actually affect any existing formulas)

**Comment by albanD** [comment]
For better or worst, the following are equivalent from the point of view of autograd:
- no graph (output requires_grad = False)
- gradient full of 0s
- gradient None (or undefined in c++)
- an error (when using autograd.grad with allow_unused=False)

In this case, I think generating 0s is easier mainly because this is an inplace function and there might be other views. If we special cased this to no require gradients, we would need special handling for view which I don't really want to think about.

That being said, we could make this more efficient by changing the formula to `self: Tensor()` as the undefined Tensor should be the same as a Tensor full of 0s.
